### PR TITLE
tls: add setup and validation callbacks for keywords

### PR DESCRIPTION
**Pull request by thus**
Add setup and validation callbacks for the following keywords:
- tls_cert_serial (check for delimiters and nocase) - uppercase the content
- tls_cert_fingerprint (check length, delimiters and nocase) - lowercase the content
- ja3_hash (check length and nocase)

Prscript:
- PR thus-pcap: https://buildbot.openinfosecfoundation.org/builders/thus-pcap/builds/154
- PR thus: https://buildbot.openinfosecfoundation.org/builders/thus/builds/154

### CODE REVIEW

File: src/detect-tls-cert-serial.c [hunk 1]
```diff
@@ -134,6 +144,74 @@ static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
...
+
+        uint32_t u;
+        for (u = 0; u < cd->content_len; u++)
+            if (cd->content[u] == ':')
+                return TRUE;
+
+        *sigerror = "No colon delimiters ':' detected in content after "
+                    "tls_cert_serial. This rule will therefore never "
+                    "match.";
+        SCLogWarning(SC_WARN_POOR_RULE, "rule %u: %s", s->id, *sigerror);
```
Comment (victorjulien) [inline src/detect-tls-cert-serial.c]: this is a warning but this leads to a rule loading error, correct? if so it should not be a warning but an error

File: src/detect-tls-cert-serial.c [hunk 2]
```diff
@@ -134,6 +144,74 @@ static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
...
+static _Bool DetectTlsSerialValidateCallback(const Signature *s,
+                                             const char **sigerror)
+{
+    const SigMatch *sm = s->init_data->smlists[g_tls_cert_serial_buffer_id];
+    for ( ; sm != NULL; sm = sm->next)
+    {
+        if (sm->type != DETECT_CONTENT)
+            continue;
+
+        DetectContentData *cd = (DetectContentData *)sm->ctx;
```
Comment (victorjulien) [inline src/detect-tls-cert-serial.c]: const

File: src/detect-tls-cert-fingerprint.c [hunk 3]
```diff
@@ -135,6 +145,83 @@ static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
...
+static _Bool DetectTlsFingerprintValidateCallback(const Signature *s,
+                                                  const char **sigerror)
+{
+    const SigMatch *sm = s->init_data->smlists[g_tls_cert_fingerprint_buffer_id];
+    for ( ; sm != NULL; sm = sm->next)
+    {
+        if (sm->type != DETECT_CONTENT)
+            continue;
+
+        DetectContentData *cd = (DetectContentData *)sm->ctx;
```
Comment (victorjulien) [inline src/detect-tls-cert-fingerprint.c]: const

File: src/detect-tls-cert-fingerprint.c [hunk 4]
```diff
@@ -135,6 +145,83 @@ static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
...
+
+        if (cd->content_len != 59) {
+            *sigerror = "Invalid length of the specified fingerprint. "
+                        "This rule will therefore never match.";
+            SCLogWarning(SC_WARN_POOR_RULE, "rule %u: %s", s->id, *sigerror);
```
Comment (victorjulien) [inline src/detect-tls-cert-fingerprint.c]: warning issued in case of error?

File: src/detect-tls-cert-serial.c (hunk 1 above)
Comment (thus) [reply src/detect-tls-cert-serial.c]: No, it's just a warning, it doesn't lead to a rule loading error. I reckon, there's some people using 'nocase' with this keyword, so it's best to just warn for now, I think at least :)

File: src/detect-tls-cert-fingerprint.c (hunk 4 above)
Comment (thus) [reply src/detect-tls-cert-fingerprint.c]: Same as above :)

File: src/detect-tls-cert-serial.c (hunk 2 above)
Comment (thus) [reply src/detect-tls-cert-serial.c]: Fixed in #3356

File: src/detect-tls-cert-fingerprint.c (hunk 3 above)
Comment (thus) [reply src/detect-tls-cert-fingerprint.c]: Fixed in #3356

File: src/detect-tls-ja3-hash.c [hunk 5]
```diff
@@ -140,6 +150,67 @@ static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
...
+static _Bool DetectTlsJa3HashValidateCallback(const Signature *s,
+                                              const char **sigerror)
+{
+    const SigMatch *sm = s->init_data->smlists[g_tls_ja3_hash_buffer_id];
+    for ( ; sm != NULL; sm = sm->next)
+    {
+        if (sm->type != DETECT_CONTENT)
+            continue;
+
+        DetectContentData *cd = (DetectContentData *)sm->ctx;
```
Comment (thus) [inline src/detect-tls-ja3-hash.c]: Also changed this to const

**Comment by thus** [comment]
Replaced by #3356

**Pull request closed without merging**